[WIP] docs: RHIDP-1486 Adding Google as an authentication provider (#85)
redhat-developer/red-hat-developers-documentation-rhdh: modules/auth/proc-adding-google-as-an-authentication-provider.adoc

`auth/userinfo.profile`::: See your personal info, including any personal info you've made publicly available.
*User type*:: Add yourself as a test user, if using the `External user` type.
.. Go to to link:https://console.cloud.google.com/apis/credentials[*APIs & Services > Credentials*], click *Create Credentials*, choose *OAuth client ID*, and create a credential with following parameters:
*Application type*:: Select `Web Application`.
*Name*:: Enter your application name, such as `{product-short}`.
*Authorized JavaScript origins*:: Add URI: `pass:[http://localhost:3000]`
*Authorized redirect URIs*:: Click *Add URI* to enter the backend authentication URI set in {product-short}:
For local development::: `pass:[http://localhost:7007/api/auth/google/handler/frame]`
For a non-local deployment::: `pass:[http://{APP_FQDN}/api/auth/google/handler/frame]`
.. Save the *Client ID* and *Client secret* for the next step.

. Set up the Google authentication provider in the {product-short} configuration (`app-config.yaml`).
+
Add the `google` provider configuration to your `app-config.yaml` under the root `auth` configuration, and optionally, enable the `google` provider for sign-in:
+

> .Sample `app-config.yaml` excerpt:

Author: ```suggestion
.Sample `app-config.yaml` fragment:
```

... In the *Application type* dropdown, select `Web Application`.
... In the *Authorized JavaScript origins* section, enter the {product-short} URI: `pass:c,a,q[https://_<APP_FQDN>_]`.
... In the *Authorized redirect URIs* section, enter the backend authentication URI set in {product-short}:
`pass:c,a,q[https://_<APP_FQDN>_/api/auth/google/handler/frame]`.
... Save the *Client ID* and *Client secret* for the next step.

. Add your Google credentials in your {product-short} secrets.
+
.. Edit your {product-short} secrets, such as `secrets-rhdh`.
+
.. Add the following key/value pairs:
+
* `AUTH_GOOGLE_CLIENT_ID`is the *Client ID* that you generated on Google.
* `AUTH_GOOGLE_CLIENT_SECRET`is the *Client secret* that you generated on Google.

. Set up the Google authentication provider in your {product-short} custom configuration.
+
.. Edit your custom {product-short} config map, such as `app-config-rhdh`.
+
.. In the `app-config-rhdh.yaml` content, add the `google` provider configuration to your `app-config.yaml` under the root `auth` configuration, and enable the `google` provider for sign-in. For example:
+
[source,yaml]
----
auth:

>   environment: development

Reviewer: ```suggestion
  environment: production
```

  providers:
    google:

>       development:

Reviewer: ```suggestion
      production:
```

[id='proc-adding-google-as-an-authentication-provider_{context}']
= Enabling the Google authentication provider

{product} includes a Google authentication provider that can authenticate users by using Google OAuth2.

.Procedure
. To allow {product-short} to authenticate with Google, create OAuth credentials in Google.
.. Log in to the link:https://console.cloud.google.com/[Google Console].
.. Select or create a new project from the dropdown menu on the top bar.
.. Go to link:https://console.cloud.google.com/apis/credentials/consent[*APIs & Services > OAuth consent screen*], and click *Edit app* to configure the consent screen:
**Authorized domains**::

> Enter your {product-short} domain (_<APP_FQDN>_).

Author: FQDN = fully qualified domain is correct.

> .. To authorize {product-short} to access the user consent screen, enter the necessary information on the https://console.cloud.google.com/apis/credentials/consent/edit[*Google Cloud > Oauth consent screen > Edit app registration*] page, including the following:

Reviewer 2: ```suggestion
.. To authorize {product-short} to access the user consent screen, enter the necessary information on the https://console.cloud.google.com/apis/credentials/consent/edit[*Edit app registration*] page, including the following:
```

Reviewer 2: The name of the page is **Edit app registration**. Adding the nav path make the name page incorrect and is also unnecessary since we link directly to the page.

... On the *Scopes* tab, select the following scopes:
+

> openid:: Associate you with your personal info on Google.
> auth/userinfo.email:: See your primary Google Account email address.
> auth/userinfo.profile:: See your personal info, including any personal info you've made publicly available.
> ... On the *Test users* tab, add yourself as a test user, if using the External user type.

Reviewer 2: ```suggestion
* *openid* associates you with your personal info on Google.
* *auth/userinfo.email* shows your primary Google Account email address.
* *auth/userinfo.profile* shows your personal info, including any personal info you've made publicly available.
... On the *Test users* tab, add yourself as a test user, if using the External user type.
```

Reviewer 2: This follows the approved "Simple list" format for definition lists per the CCS style guide

*Application type*:: Select `Web Application`.
*Name*:: Enter your application name, such as `{product-short}`.
*Authorized JavaScript origins*:: Add URI: `pass:[{APP_FQDN}]`
*Authorized redirect URIs*:: Click *Add URI* to enter the backend authentication URI set in {product-short}:

> `pass:[{APP_FQDN}/api/auth/google/handler/frame]`

Author: ```suggestion
`pass:c,a,q[_<APP_FQDN>_/api/auth/google/handler/frame]`
```

.. Save the *Client ID* and *Client secret* for the next step.

. Add your Google credentials in your {product-short} secrets.
+
Edit your {product-short} secrets, such as `secrets-rhdh`.
+
Add the following key/value pairs:
+
`AUTH_GOOGLE_CLIENT_ID`:: Enter the *Client ID* that you generated on Google.
`AUTH_GOOGLE_CLIENT_SECRET`:: Enter the *Client secret* that you generated on Google.

> 

Reviewer 2: Numbers are missing from these steps. If they are sub-steps, format them as such.

Author: I mean: intentionally unordered.

Reviewer 2: So _Edit your {product-short} secrets, such as `secrets-rhdh`_ and _Add the following key/value pairs_ are not procedural steps? What is being defined there?

> . Add your Google credentials in your {product-short} secrets.
> +
> Edit your {product-short} secrets, such as `secrets-rhdh`.
> +
> Add the following key/value pairs:
> +
> `AUTH_GOOGLE_CLIENT_ID`:: Enter the *Client ID* that you generated on Google.
> `AUTH_GOOGLE_CLIENT_SECRET`:: Enter the *Client secret* that you generated on Google.

Reviewer 2: ```suggestion
. Add your Google credentials in your {product-short} secrets.
+
.. Edit your {product-short} secrets, such as `secrets-rhdh`.
+
.. Add the following key/value pairs:
+
* `AUTH_GOOGLE_CLIENT_ID`is the *Client ID* that you generated on Google.
* `AUTH_GOOGLE_CLIENT_SECRET`is the *Client secret* that you generated on Google.
```

. Set up the Google authentication provider in your {product-short} custom configuration.
+
Edit your custom {product-short} config map, such as `app-config-rhdh`.
+

> In the `app-config-rhdh.yml` content, add the `google` provider configuration to your `app-config.yaml` under the root `auth` configuration, and enable the `google` provider for sign-in:

Author: ```suggestion
In the `app-config-rhdh.yaml` content, add the `google` provider configuration to your `app-config.yaml` under the root `auth` configuration, and enable the `google` provider for sign-in:
```

+

> .Sample `app-config.yaml` fragment:

Author: ```suggestion
.Sample `app-config-rhdh.yaml` fragment:
```

> . Set up the Google authentication provider in your {product-short} custom configuration.

Reviewer 3: @themr0c @linfraze After configuring the Google Auth provider in app-config, logging in might result in a "user not found" error. To resolve this, the following configuration is required:

app-config.yaml
```
catalog:
      locations:
        - type: url
          target: https://github.com/janus-qe/rhdh-test/blob/main/user.yml
          rules:
            - allow: [User]
```
user.yml
```
apiVersion: backstage.io/v1alpha1
kind: User
metadata:
  name: rhdhtest
spec:
  memberOf: []
```

Here, name: rhdhtest is user ID of the Google account.

Reviewer 2: @subhashkhileri thanks for letting us know! Is this step required or is it a troubleshooting step? If not required to complete the task, it might be better suited for a troubleshooting sub-section. If that is the case, we can address this with a separate Jira ticket and determine the best format and placement for the troubleshooting steps.

Reviewer 3: @linfraze This step is required.

Reviewer 2: @subhashkhileri The part about _logging in might result in a "user not found" error. To resolve this, the following configuration is required_? Seemed like that configuration is only required if there is an error, which would make it a troubleshooting step. But if this login error will always occur and the step is truly required to prevent it, then we can incorporate this into the main procedure.

Reviewer 3: @Zaperex  can you please confirm if its always required or not?

Author: I am not sure how to write it to make it scale for enterprise customers who have more than one user. 

How to declare a user list rather than just one user?

Is it possible to declare the user list as a Kubernetes object rather than using a file on GitHub?

Author: @kadel Do you have any hint?

Reviewer: This is actually how backstage works. User that are logging-in needs to have an existing entity in catalog. But this is not specific to Google auth, the same applies for any auth provider. Maybe we should mention that somewhere in generic auth section. 

Users are regular catalog entities, so they can be imported from any file, or there are even plugins that can import users from other systems like GitHub Orgs or Keycloak. I'm not aware of any importer plugins for Google.

The way how authenticated users are mapped to catalog users can by customized by sign-in resolvers (https://backstage.io/docs/auth/identity-resolver#sign-in-resolvers)

Author: I understand we have a prerequisite, that is the same for every provider, and that we could document in another place, is that correct:

In this doc in the Prerequsites section:

```

* You have provisioned all users in the {product} catalog.

```

In another doc:

```
# Provisioning users in the {product} catalog

Signing in a user into {product} requires a mapping of the user identity from the third-party auth provider to a {product-short} user identity. 

A {product-short} user identity consists in: 
- A user entity reference.
- A set of ownership references.

.Procedure

. Define the user entiry reference. TBD

. Define set of ownership references. TBD

.Verification

. Get the user list. TBD
. Get the ownership reference. TBD

```

Did I understand well?

> .Sample `app-config-rhdh.yaml` fragment:

Author: ```suggestion
.`app-config-rhdh.yaml` fragment:
```

> In the `app-config-rhdh.yaml` content, add the `google` provider configuration to your `app-config.yaml` under the root `auth` configuration, and enable the `google` provider for sign-in:

Reviewer 2: These are the steps that seem to be missing numbers. And the key/value pairs might need bullets.

[source,yaml]
----
auth:
  environment: development
  providers:
    google:
      development:
        clientId: ${AUTH_GOOGLE_CLIENT_ID}
        clientSecret: ${AUTH_GOOGLE_CLIENT_SECRET}
 signInPage: google # <1>
----

<1> To enable the Google provider as default sign-in provider.

> 

Reviewer 2: Could some of these be separate procedure modules within a Google Authentication Provider assembly? See the GitHub Authentication Provider assembly as an example.

Author: I don't think so.

Steps are referencing to each other (Client ID and Client Secret), and don't make sense in isolation.

Splitting in different procedures would require adding a lot of overhead (prerequisites for the previous step being completed), in contradiction with minimalism.

Reviewer 2: If all these steps are truly part of one procedure, then we can just fix the formatting so that all steps are numbered properly.

> . Set up the Google authentication provider in your {product-short} custom configuration.
> +
> Edit your custom {product-short} config map, such as `app-config-rhdh`.
> +
> In the `app-config-rhdh.yaml` content, add the `google` provider configuration to your `app-config.yaml` under the root `auth` configuration, and enable the `google` provider for sign-in:
> +
> .`app-config-rhdh.yaml` fragment:
> [source,yaml]
> ----
> auth:
>   environment: development
>   providers:
>     google:
>       development:
>         clientId: ${AUTH_GOOGLE_CLIENT_ID}
>         clientSecret: ${AUTH_GOOGLE_CLIENT_SECRET}
>  signInPage: google # <1>
> ----
> 
> <1> To enable the Google provider as default sign-in provider.
> 

Reviewer 2: ```suggestion
. Set up the Google authentication provider in your {product-short} custom configuration.
+
.. Edit your custom {product-short} config map, such as `app-config-rhdh`.
+
.. In the `app-config-rhdh.yaml` content, add the `google` provider configuration to your `app-config.yaml` under the root `auth` configuration, and enable the `google` provider for sign-in. For example:
+
[source,yaml]
----
auth:
  environment: development
  providers:
    google:
      development:
        clientId: ${AUTH_GOOGLE_CLIENT_ID}
        clientSecret: ${AUTH_GOOGLE_CLIENT_SECRET}
 signInPage: google # <1>
----

<1> Enables the Google provider as default sign-in provider.

```

and click *Edit app* to configure the consent screen:
**Authorized domains**::
For local development:: You do not need to enter any Authorized domain.
For a non-local deployment::: Enter your {product} domain.
*Scopes*:: Add these scopes:
`openid`::: Associate you with your personal info on Google.
`auth/userinfo.email`::: See your primary Google Account email address.
`auth/userinfo.profile`::: See your personal info, including any personal info you've made publicly available.
*User type*:: Add yourself as a test user, if using the `External user` type.
.. Go to to link:https://console.cloud.google.com/apis/credentials[*APIs & Services > Credentials*],
click *Create Credentials*, choose *OAuth client ID*, and create a credential with following parameters:
*Application type*:: Select `Web Application`.
*Name*:: Enter your application name, such as `{product}`.
*Authorized JavaScript origins*:: Add URI: `pass:[http://localhost:3000]`
*Authorized redirect URIs*:: Click *Add URI* to enter the backend authentication URI set in {product}:
For local development::: `pass:[http://localhost:7007/api/auth/google/handler/frame]`

> For a non-local deployment::: `pass:[http://{APP_FQDN}:{APP_BACKEND_PORT}/api/auth/google/handler/frame]`

Reviewer: for non local deployment it will be only http://{rhdh_insteance_fqdn}/api/auth/google/handler/frame

btw, I've noticed that we are using different placeholder for RHDH the hostname of the RHDH instance that is getting configured. It would be nice if we could unify that

Reviewer 4: we can add more reusable strings into https://github.com/redhat-developer/red-hat-developers-documentation-rhdh/blob/main/artifacts/attributes.adoc to make life easier.

note that if you're then going to use those variables in a code snippet you need `subs="attributes"` in your code block. See examples:

https://github.com/redhat-developer/red-hat-developers-documentation-rhdh/pull/99/files#diff-0f7e8a8fb3fdb835a98a8a937c0c87986260380b80b77a1f84f8ca529be430a2

Author: ```suggestion
For a non-local deployment::: `pass:[http://{APP_FQDN}/api/auth/google/handler/frame]`
```

The app renders this component before any other routes.
The component provides the identity of the current user.
The `SignInPage` can render any number of pages and components, or just blank space with logic running in the background.
In the end however it must provide a valid {product} user identity through the onSignInSuccess callback prop, at which point the rest of the app is rendered.
+
If you want to, you can use the `SignInPage` component provided by `@backstage/core-components`, which takes either a provider or providers (array) prop of `SignInProviderConfig` definitions.
+
The following example for Google shows the additions needed to `packages/app/src/App.tsx`:
+
.`packages/app/src/App.tsx`
[source,javascript]
----
import { googleAuthApiRef } from '@backstage/core-plugin-api';
import { SignInPage } from '@backstage/core-components';
const app = createApp({
  components: {
    SignInPage: props => (
      <SignInPage
        {...props}
        auto
        provider={{
          id: 'google-auth-provider',
          title: 'Google',
          message: 'Sign in using Google',
          apiRef: googleAuthApiRef,
        }}
      />
    ),
  },
  // ..
});

> ----

Reviewer: This whole section is not necessary. in RHDH we have `signInPage` option that simplifies all this.  So all that is required is to add `signInPage: google` to the configuration. 

This is partially explained in https://github.com/redhat-developer/red-hat-developers-documentation-rhdh/pull/76 But only partially, we need better documentation of what this option is for.
